cpp/include/raft/matrix/col_wise_sort.cuh

```diff
@@ -50,7 +51,105 @@ void sort_cols_per_row(const InType* in,
+ */
+template <typename in_t, typename out_t, typename matrix_idx_t>
+void sort_cols_per_row(const raft::handle_t& handle,
+                       raft::device_matrix_view<const in_t, matrix_idx_t, raft::row_major> in,
+                       raft::device_matrix_view<out_t, matrix_idx_t, raft::row_major> out,
+                       std::optional<raft::device_matrix_view<in_t, matrix_idx_t, raft::row_major>>
+                         sorted_keys = std::nullopt)
+{
+  RAFT_EXPECTS(in.extent(1) == out.extent(1) && in.extent(0) == out.extent(0),
+               "Input and output matrices must have the same shape.");
+
+  if (sorted_keys.has_value()) {
+    RAFT_EXPECTS(in.extent(1) == sorted_keys.value().extent(1) &&
+                   in.extent(0) == sorted_keys.value().extent(0),
+                 "Input and `sorted_keys` matrices must have the same shape.");
+  }
+
+  size_t workspace_size = 0;
+  bool alloc_workspace  = false;
+
+  in_t* keys = sorted_keys.has_value() ? sorted_keys.value().data_handle() : nullptr;
+
+  detail::sortColumnsPerRow<in_t, out_t>(in.data_handle(),
+                                         out.data_handle(),
+                                         in.extent(0),
+                                         in.extent(1),
+                                         alloc_workspace,
+                                         (void*)nullptr,
+                                         workspace_size,
+                                         handle.get_stream(),
+                                         keys);
+
+  if (alloc_workspace) {
+    auto workspace = raft::make_device_vector<char>(handle, workspace_size);
+
+    detail::sortColumnsPerRow<in_t, out_t>(in.data_handle(),
+                                           out.data_handle(),
+                                           in.extent(0),
+                                           in.extent(1),
+                                           alloc_workspace,
+                                           (void*)workspace.data_handle(),
+                                           workspace_size,
+                                           handle.get_stream(),
+                                           keys);
+  }
+}
+
+namespace sort_cols_per_row_impl {
+template <typename T>
+struct sorted_keys_alias {
+};
+
+template <>
+struct sorted_keys_alias<std::nullopt_t> {
+  using type = double;
+};
+
+template <typename in_t, typename matrix_idx_t>
+struct sorted_keys_alias<
+  std::optional<raft::device_matrix_view<in_t, matrix_idx_t, raft::row_major>>> {
+  using type = typename raft::device_matrix_view<in_t, matrix_idx_t, raft::row_major>::value_type;
+};
+
+template <typename T>
+using sorted_keys_t = typename sorted_keys_alias<T>::type;
+}  // namespace sort_cols_per_row_impl
+
+/**
+ * @brief Overload of `sort_keys_per_row` to help the
+ *   compiler find the above overload, in case users pass in
+ *   `std::nullopt` for one or both of the optional arguments.
```

Comment:
I thought this overload was only needed when there are two optional arguments? @mhoemmen